clippy_lints/src/direct_recursion.rs

```diff
@@ -0,0 +1,245 @@
+    "functions shall not call themselves directly"
+}
+
+pub struct DirectRecursion {
+    fn_id_stack: Vec<LocalDefId>,
+    expr_check_blocker: Option<HirId>,
+}
+
+impl DirectRecursion {
+    pub fn new(_: &'static Conf) -> Self {
+        Self {
+            // We preallocate a stack of size 4, because we'll probably need more than 2
+            // but I really don't expect us to ever see more than 4 nested functions
+            fn_id_stack: Vec::with_capacity(4),
+            expr_check_blocker: None,
+        }
+    }
+
+    /// Blocks checking more expressions, using `expr` as the key.
+    fn block_with_expr(&mut self, expr: &Expr<'_>) {
+        self.expr_check_blocker = Some(expr.hir_id);
+    }
+
+    /// Tells whether we're currently allowed to check expressions or not
+    fn is_blocked(&self) -> bool {
+        self.expr_check_blocker.is_some()
+    }
+
+    /// Tries opening the lock using `expr` as the key.
+    fn try_unlocking_with(&mut self, expr: &Expr<'_>) {
+        if let Some(key) = self.expr_check_blocker
+            && key == expr.hir_id
+        {
+            self.expr_check_blocker = None;
+        }
+    }
+}
+
+impl_lint_pass!(DirectRecursion => [DIRECT_RECURSION]);
+
+impl<'tcx> LateLintPass<'tcx> for DirectRecursion {
+    /// Whenever we enter a Body, we push its owner's `DefId` into the stack
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &Body<'_>) {
+        self.fn_id_stack.push(cx.tcx.hir_body_owner_def_id(body.id()));
+    }
+
+    /// We then revert this when we exit said `Body`
+    fn check_body_post(&mut self, _: &LateContext<'tcx>, _: &Body<'_>) {
+        _ = self.fn_id_stack.pop();
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        // We use this inner lock so that we avoid recursing if we've
+        // already linted an expression that contains the expression
+        // we're now visiting.
+        // This lock is closed whenever we emit a lint, and it's opened
+        // after we exit the node that closed it (see `check_expr_post`)
+        if self.is_blocked() {
+            return;
+        }
+
+        // Before running the lint, we look up the attributes of this Expr.
+        // If it has been marked with `clippy::allowed_recursion`, then
+        // we ignore it, as well as everything inside it; someone has
+        // decided that the recursive calls within it are fine.
+        let attrs = cx.tcx.hir_attrs(expr.hir_id);
+        if get_attr(cx.sess(), attrs, sym::allowed_recursion).next().is_some() {
+            self.block_with_expr(expr);
+            return;
+        }
+
+        match expr.kind {
+            ExprKind::MethodCall(_, _, _, _) => {
+                // Uniquely identifying the `DefId` of method calls requires doing type checking.
+                // `cx` takes care of that, and then we can ask for the `type_dependent_def_id`
+                // of the `expr` whose kind is `ExprKind::MethodCall`.
```

Comment:
You don't need to explain every step of what the lint is doing here, `type_dependent_def_id()` is used everywhere throughout Clippy (100+ times), particularly on method calls, it is known what it is doing.